proposal.rst

```diff
@@ -1242,6 +1242,23 @@ components, thus freeing both server and application authors from
 implementing the same features over and over again.
 
 
+HTTP/2 Features
+---------------
+
+From the perspective of WSGI applications and servers, HTTP/2 is not handled
+differently from HTTP/1.1. Because the semantics of HTTP/2 requests and
+responses are identical to the semantics of HTTP/1.1 requests and responses,
+there is no requirement for servers to treat them differently in WSGI.
+
+Servers that support both HTTP/1.1 and HTTP/2 **may** provide support for
```

Comment:
The sever *does* have access to the Link information in the only way that matters.

HTTP/2 push works by constructing a "fake" request that is used to correlate with a single response. The way this would work in this case is that that application would send the Link header, which the server would catch. The server would generate this fake request, send it *back* to the application, then push the application's response.

Let me sketch this out below:

1. The server receives the following request and passes it to the application in the standard WSGI manner (needing to construct the WSGI environ dict and so on):

    ```
    :path: /index.html
    :authority: example.com
    :scheme: https
    :method: GET
    host: example.com
    accept: */*
    ```

2. The application sends a 200 response that contains the header `Link: </static/style.css>; rel=preload; as=style`.
3. The server catches that header. It supports HTTP/2 server push by this extension, and so it immediately generates a new "fake" request for that resource, which looks like this:

    ```
    :path: /static/style.css
    :authority: example.com
    :scheme: https
    :method: GET
    host: example.com
    accept: */*
    ```
4. The server then does two things with this virtual request. Firstly, it pushes it to the client in a PUSH_PROMISE frame. Secondly, it sends it to the application *exactly like* it received it from the real client: in a brand new WSGI function call.
5. The application responds as normal to the new request, despite the fact that this request was actually initiated by the server.

In this model, the complexities of pushing resources are handled entirely within the server. This allows for some neat optimisations: for example, servers that can be both WSGI servers and serve static files (such as Apache) can potentially run the pushed request through their complete request routing logic, allowing for the application to push static files directly. It also allows us to maintain the standard WSGI interface while opening the possibility for the application to provide push hints to the server.

Finally, the push hint here is also of use to the client in the case where pushing is not possible. Servers that do not opt in or do not support HTTP/2 still function correctly, and clients behave as expected.

Does that make sense?